### DISCUSSION TITLE
How to use things derived from QAbstractListModel as field/property?

### DISCUSSION THREAD

**Discussion by araraloren** [KDAB/cxx-qt#914] [Q&A] [answered]
Hi, I know we can create a type derived from QAbstractListModel, such as
```
    extern "RustQt" {

        #[qobject]
        #[base = "QAbstractListModel"]
        #[qml_element]
        type CustomBaseClass = super::CustomBaseClassRS;
    }

    unsafe extern "RustQt" {
        #[qobject]
        #[qml_element]
        #[qproperty(CustomBaseClass, files_model)]
        type FileServer = super::FileServerRS;
    }
```
Could someone tell me how can I use CustomBaseClass in the filed, so I can access it in qml file?
```
{
...
FileServer {
 id: fileServer
}

...
model: fileServer.filesModel
...
}
```
Thanks

**Comment by ahayzen-kdab** [comment, marked as answer]
Hey,

So QML will be wanting a raw pointer in C++, eg `CustomBaseClass*`. We can use Rust raw pointers for this, if we look at the `nested_qobjects` example this is similar.


This says that there is a Q_PROPERTY which is a pointer under the field inner.
https://github.com/KDAB/cxx-qt/blob/f82ab4e4ed90a99cae5aaf493a52ef29de73aee5/examples/qml_features/rust/src/nested_qobjects.rs#L32-L35

Then we store a raw pointer to the QObject as a field in the struct

https://github.com/KDAB/cxx-qt/blob/f82ab4e4ed90a99cae5aaf493a52ef29de73aee5/examples/qml_features/rust/src/nested_qobjects.rs#L67-L78

This will need to go into your FileServerRS.

Note this example also shows how to use the pointer with methods and shows how to get from a raw pointer back to a pinned mutable reference (so that you can reach methods on the pointer object.

Note that you might need a C++ method to create the raw pointer of the CustomBaseClass if you are doing this in Rust rather than QML. (we have an issue somewhere about generating helpers for this automatically somewhere i think).

Feel free to ask more questions as this area is a little tricky :-)

**Reply by araraloren** [reply]
Thanks for your reply!

**Comment by realroot2185** [comment]
https://doc.qt.io/qt-6/qabstractlistmodel.html
> When subclassing QAbstractListModel, you must provide implementations of the [rowCount](https://doc.qt.io/qt-6/qabstractitemmodel.html#rowCount)() and [data](https://doc.qt.io/qt-6/qabstractitemmodel.html#data)() functions. Well behaved models also provide a [headerData](https://doc.qt.io/qt-6/qabstractitemmodel.html#headerData)() implementation.

Is there an example to how use a type derived from QAbstractListModel for a QML ListView where you can add/remove elements and save/load the list?

I cannot call the function from the type instance in QML:
`main.qml:109: TypeError: Property 'do_something' of object MyType(0x5dbb2a6c9810) is not a function`

**Reply by ahayzen-kdab** [reply]
This example has a `QAbstractListModel` implemented in Rust and has various `add` methods that are called from QML.

https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/custom_base_class.rs

https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/qml/pages/CustomBaseClassPage.qml

**Reply by realroot2185** [reply]
Any suggestion to how save/load models in the ListView from a file?

So that when you open it there are already models from the last time.
And initially it loads the default models from a file.

**Reply by ahayzen-kdab** [reply]
So you'll want to use the Initialize trait to run code in the constructor of the QObject https://docs.rs/cxx-qt/latest/cxx_qt/trait.Initialize.html this could then be used to load things. Then probably the Drop trait on the struct might be enough to write values in the fields to a file.

**Reply by LeonMatthesKDAB** [reply]
Also, take a look at this example: https://github.com/KDAB/cxx-qt/blob/main/examples/qml_features/rust/src/serialisation.rs

Which shows how to serialize/deserialize data stored in a QObject.

@ahayzen-kdab By the way, I think we should be able to simplify this example, a lot.
I don't think we need the separate data struct anymore with our current API :thinking: 
We can just derive Serialize and Deserialize on the inner struct itself.
